[DiagnosticBridge] Make sure that diagnostic queues up its child notes

xedin opened the pull request:
`DiagnosticEngine` has an API that allows to attach notes to a "primary" diagnostic (an error or a warning). This works well with old formatting (`llvm`) but `swift` formatter doesn't display attached notes which makes some diagnostics very hard to work with i.e. `invalid_redecl`.

<!--
If this pull request is targeting a release branch, please fill out the
following form:
https://github.com/swiftlang/.github/blob/main/PULL_REQUEST_TEMPLATE/release.md?plain=1

Otherwise, replace this comment with a description of your changes and
rationale. Provide links to external references/discussions if appropriate.
If this pull request resolves any GitHub issues, link them like so:

  Resolves <link to issue>, resolves <link to another issue>.

For more information about linking a pull request to an issue, see:
https://docs.github.com/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue
-->

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/main/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[DiagnosticBridge] Make sure that diagnostic queues up its child notes

`DiagnosticEngine` has an API that allows to attach notes to a "primary"
diagnostic (an error or a warning). This works well with old formatting
(`llvm`) but `swift` formatter doesn't display attached notes which makes
some diagnostics very hard to work with i.e. `invalid_redecl`.

xedin commented:
@swift-ci please smoke test

ktoso reviewed (approved):
Thanks a lot @xedin !

Diagnosing redeclarations was a nightmare without this in https://github.com/swiftlang/swift/pull/79600

Thanks!

xedin commented:
@cachemeifyoucan That’s why I added you, looking for suggestions for how to test this properly.

xedin committed:
[Frontend] Treat empty diagnostic category as "not present"

Serialized diagnostic stores an `std::string` which means that
the category gets marked as present even if its empty, which is
incorrect.

cachemeifyoucan commented on lib/Frontend/CachedDiagnostics.cpp:
```diff
@@ -617,7 +617,7 @@ llvm::Error DiagnosticSerializer::deserializeDiagnosticInfo(
                                   Info.FormatString,
                                   {},
-                                  Info.Category,
+                                  Info.Category.empty() ? StringRef() : Info.Category,
```
I can't see what difference it makes? Can you explain?

xedin replied:
The parameter expects an optional pointer when passed to bridging so without this we'd happy pass a valid empty string instead of `nil` and that leads to swift-syntax printing `[#]` as a category because the expectation is that a valid category always has a name.

xedin replied:
That was added here - https://github.com/swiftlang/swift-syntax/pull/2981/files#diff-0a73410f2b8e6bd296483c34ce6f07228312dffe3eaafc520ef4bae5f0cbe157R54

cachemeifyoucan replied:
Is this because the `StringRef` constructed from an empty string doesn't have `data()` to be `nullptr`?

This is a very unexpected result and very hard to understand here. Is it possible to get `swift_ASTGen_addQueuedDiagnostic` to bridge the empty StringRef as `nil`?

cachemeifyoucan replied:
Or `DiagnosticBridge.cpp` will pass nullptr instead? Doing it here is a bit hard to understand

xedin replied:
I'm not sure what you mean, we are using a `SerializedDiagnosticInfo` that has std::string for this field, empty std::string is not a nullptr, that's why we are in this situation.

cachemeifyoucan replied:
Right, that is what I understand, but I wonder if it should be bridged elsewhere instead.

If category can ever be empty, then the bridging in `swift_ASTGen_addQueuedDiagnostic` should just treat as zero length stringref as nil.
If category can be empty, then the abstraction in DiagnosticInfo needs to be updated, together with this serialization format.

xedin replied:
I don't think that is always the case, it's definitely is for category though, in generally I doubt that.

xedin committed:
[Tests] NFC: Add a test-case for printing child notes associated with errors/warnings

xedin commented:
@swift-ci please test

xedin committed:
[DiagnosticsBridge] Treat empty category name as `nil`

Default value for `Category` for serialization purposes is
an empty string but it should be handled as `nil` while
bridging because category name cannot be empty when present.

xedin commented:
@swift-ci please test

cachemeifyoucan commented:
LGTM

bnbarham commented on test/diagnostics/pretty-printed-diagnostics.swift:
```diff
@@ -85,3 +88,7 @@ foo(b:
 // CHECK:              |         `- error: argument 'a' must precede argument 'b'
 // CHECK: [[#LINE+1]]  |
+
+// CHECK: SOURCE_DIR{{[/\]+}}test{{[/\]+}}diagnostics{{[/\]+}}pretty-printed-diagnostics.swift:[[#LINE:]]:6
```
It'd be nice to check for the error here too, I was initially very confused by this and thought that it was somehow a note off on its own 😅

xedin replied:
✅

bnbarham commented on lib/AST/DiagnosticBridge.cpp:
```diff
@@ -77,6 +77,13 @@ static void addQueueDiagnostic(void *queuedDiagnostics,
+  // TODO: A better way to do this would be to pass the notes as an
+  // argument to `swift_ASTGen_addQueuedDiagnostic` but that requires
+  // bridging of `Note` structure and new serialization.
+  for (auto *childNote : info.ChildDiagnosticInfo) {
+    addQueueDiagnostic(queuedDiagnostics, *childNote, SM);
+  }
```
Might be clearer to move this up into `enqueueDiagnostic`?

xedin replied:
I'd prefer to keep this in here for now since this is technically an implementation detail a "primary" diagnostic and ideally should be dispatched as an argument like what I mentioned in a TODO.

bnbarham replied:
Unless I'm misunderstanding, the infrastructure currently assumes that notes are attached to the first primary diagnostic in the group. Ie. `enqueueDiagnostic` isn't necessarily a primary diagnostic anyway. But 🤷‍♂️

xedin replied:
You might be talking about lower level things but there is a dedicated API in DiagnosticEngine which let's the users attach notes directly to the diagnostic they want. I think we should replace all of the disjoint emission of notes with `diagnoseWithNotes` and remove any inference that happens later on...

bnbarham replied:
Yeah, all for moving to `diagnoseWithNotes`. Happy to leave it where it is 👍

bnbarham reviewed (commented):
Thanks for fixing 🙇‍♂️

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test Linux platform
